core/src/main/scala/kafka/raft/KafkaRequestPurgatory.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.raft
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{DelayedOperation, DelayedOperationPurgatory}
+import kafka.utils.Logging
+import kafka.utils.timer.Timer
+import org.apache.kafka.common.errors.TimeoutException
+import org.apache.kafka.raft.RequestPurgatory
+
+/**
+ * Simple purgatory shim for integration with the Raft library. We assume that
+ * both [[await()]] and [[completeAll()]] are called in the same thread.
+ */
+class KafkaRequestPurgatory(brokerId: Int, timer: Timer, reaperEnabled: Boolean = true)
+  extends RequestPurgatory[Unit] with Logging {
+
+  private val key = new Object()
+  private val purgatory = new  DelayedOperationPurgatory[DelayedRaftRequest](
+    "raft-request-purgatory", timer, brokerId, reaperEnabled = reaperEnabled)
+
+  override def await(future: CompletableFuture[Unit], maxWaitTimeMs: Long): Unit = {
```

Comment:
If no HW advance but many fetches are coming from both voters and observers, will that invalidate purgatory's purpose as we always check for unfinished delayed operation?

Reply:
I'm not sure I follow the question. I think we only try to complete the new operation here, so any other requests are just stuck in purgatory until their timeout expires or the high watermark advances.